src/linear_algebra/matrix/transvection.lean

```diff
@@ -0,0 +1,754 @@
+/-!
+# Transvections
+
+Transvections are matrices of the form `1 + c E i j`, where `E i j` is the elementary matrix with
+a `1` at position `(i, j)`. Multiplying by such a transvection on the left (resp. on the right)
+amounts to adding `c` times the `j`-th row to to the `i`-th row (resp `c` times the `i`-th column
+to the `j`-th column). Therefore, they are useful to present algorithms operating on lines and
+columns.
+
+We show that, over a field, any matrix can be written as `L ⬝ D ⬝ L'`, where `L` and `L'` are
+product of transvections and `D` is diagonal. In other words, one can reduce a matrix to diagonal
+form by operations on its rows and columns, a variant of Gauss' pivot algorithm.
+
+## Main definitions and results
+
+* `E R i j` is the elementary matrix with a one at position `(i, j)`.
+* `transvection i j c` is the matrix equal to `1 + c • E R i j`.
+* `transvection_struct n R` is a structure containing the data of `i, j, c` and a proof that
+  `i ≠ j`. These are often easier to manipulate than straight matrices, especially in inductive
+  arguments.
+
+* `exists_list_transvec_mul_diagonal_mul_list_transvec` states that any matrix `M` over a field can
+  be written in the form `t_1 ⬝ ... ⬝ t_k ⬝ D ⬝ t'_1 ⬝ ... ⬝ t'_l`, where `D` is diagonal and
+  the `t_i`, `t'_j` are transvections.
+
+* `diagonal_transvection_induction` shows that a property which is true for diagonal matrices and
+  transvections, and invariant under product, is true for all matrices.
+* `diagonal_transvection_induction_of_det_ne_zero` is the same statement over invertible matrices.
+
+## Implementation details
+
+The proof of the reduction results is done inductively on the size of the matrices, reducing an
+`(r + 1) × (r + 1)` matrix to a matrix whose last row and column are zeroes, except possibly for
+the last diagonal entry. This step is done as follows.
+
+If all the coefficients on the last row and column are zero, there is nothing to do. Otherwise,
+one can put a nonzero coefficient in the last diagonal entry by a row or column operation, and then
+subtract this last diagonal entry from the other entries in the last row and column to make them
+vanish.
+
+This step is done in the type `fin r ⊕ unit`, where `fin r` is useful to choose arbitrarily some
+order in which we cancel the coefficients, and the sum structure is useful to use the formalism of
+block matrices.
+
+To proceed with the induction, we reindex our matrices to reduce to the above situation.
+-/
+
+universes u₁ u₂
+
+namespace matrix
+open_locale matrix
+
+variables (n p : Type*) (R : Type u₂) {𝕜 : Type*} [field 𝕜]
+variables [decidable_eq n] [decidable_eq p]
+variables [comm_ring R]
+
+section elementary_basis
+
+variables {n} (i j : n)
+
+/-- The elementary matrix `E i j` has a one at position `(i, j)`, and zeroes elsewhere.
+It is useful to define these matrices for explicit calculations in sl n R. -/
+@[reducible] definition E : matrix n n R := λ i' j', if i = i' ∧ j = j' then 1 else 0
```

Comment:
What's the advantage of making this reducible? How does this definition compare?
```suggestion
def E : matrix n n R
| i' j' := if i = i' ∧ j = j' then 1 else 0
```
Which is a little more similar to `transpose`.

Reply:
Also, presumably there is no need to restrict this to square matrices?

Reply:
All the part about `E` is not from this PR, just moved from the file `lie/classical.lean`.

Reply:
I somewhat feel we should just remove this, or at least use:
```suggestion
@[reducible] definition E : matrix n n R := matrix.std_basis_matrix i j 1
```